### PR TITLE
test: expand the pixel tests section

### PR DESCRIPTION
**Pull request by jelly**
Explain the pixel tests architecture, add examples and include some pixel test contributor notes.

### CODE REVIEW

File: test/README.md [hunk 1]
```diff
@@ -84,10 +84,70 @@ ssh and web.  See the "Helpful tips" section below.
 
 ## Pixel tests
 
-The verify test suite contains ["pixel tests"](https://cockpit-project.org/blog/pixel-testing.html).
-Make sure to create the test/reference submodule before running tests which contain pixel tests.
+Pixel tests in Cockpit ensure that updates of our dependencies or code changes
+don't break unexpectedly break the UI, for example slight changes of layout,
```
Comment (martinpitt) [inline test/README.md:88]:
```suggestion
don't break the UI: for example slight changes of layout,
```

File: test/README.md [hunk 2]
```diff
@@ -84,10 +84,70 @@ ssh and web.  See the "Helpful tips" section below.
...
+padding, color and everything which isn't easily tested by a human. They also
```
Comment (martinpitt) [inline test/README.md:89]:
```suggestion
padding, color and everything which isn't easily spotted by a human. They also
```

File: test/README.md [hunk 3]
```diff
@@ -84,10 +84,70 @@ ssh and web.  See the "Helpful tips" section below.
...
+give us confidence that an update of our UI Framework doesn't introduce changes
+in how Cockpit looks.
+
+Pixel tests are in essence quite simple, they make a screenshot of the current
+state and compare this to a known good state, if there is a difference the test
+fails and a pixel difference is shown. This works as our tests run in the
```
Comment (martinpitt) [inline test/README.md:95]:
```suggestion
Pixel tests make a screenshot of a selector and compare it to a known good 
reference image. if there is a difference, the test fails and a pixel difference
is shown.

This works as our tests run in the
```

File: test/README.md [hunk 4]
```diff
@@ -84,10 +84,70 @@ ssh and web.  See the "Helpful tips" section below.
...
+[cockpit/tasks container](https://quay.io/repository/cockpit/tasks) which pins
+the browser and font rendering so repeated runs provide the same pixels. To
+generate new pixels this tasks container should be used as your own browser and
```
Comment (martinpitt) [inline test/README.md:98]:
```suggestion
generate new pixels, this tasks container must be used; your own browser and
```

File: test/README.md [hunk 5]
```diff
@@ -84,10 +84,70 @@ ssh and web.  See the "Helpful tips" section below.
...
+
+The test images are stored in git submodule in the `test/reference` directory and be fetched with:
+
+```sh
+./test/common/pixel-tests fetch
+```
+
+As Cockpit tests under multiple distributions and it is not worth the effort to
+run pixel tests on every supported distribution we only run them for the
+configured image configured in `test/reference-image`.
```
Comment (martinpitt) [inline test/README.md:110]:
```suggestion
image configured in `test/reference-image`.
```

File: test/README.md [hunk 6]
```diff
@@ -84,10 +84,70 @@ ssh and web.  See the "Helpful tips" section below.
...
+
+Our tests don't know when it is interesting to make a screenshot of the current
+state, so our tests have to specifically assert for saved pixels using
+`browser.assert_pixels`. This assertion method requires an CSS selector to be
```
Comment (martinpitt) [inline test/README.md:114]:
```suggestion
Our tests call `Browser.assert_pixels` at interesting and strategic places. This assertion
method requires at least a CSS selector and an image title.
```

File: test/README.md [hunk 7]
```diff
@@ -84,10 +84,70 @@ ssh and web.  See the "Helpful tips" section below.
...
+provided and a title of the image. Pixel tests are generated in five layouts by
```
Comment (martinpitt) [inline test/README.md:115]:
```suggestion
Pixel tests are generated in five layouts by
```

File: test/README.md [hunk 8]
```diff
@@ -84,10 +84,70 @@ ssh and web.  See the "Helpful tips" section below.
...
+default; desktop, medium, mobile, dark and rtl.
+
+Take a screenshot the content in `#detail-content`:
```
Comment (martinpitt) [inline test/README.md:118]:
```suggestion
Take a screenshot of the content in `#detail-content`:
```

File: test/README.md [hunk 9]
```diff
@@ -84,10 +84,70 @@ ssh and web.  See the "Helpful tips" section below.
...
+```
+
+Take a screenshot the content in `#detail-content` and skip it for a specific
+layout as it generates unstable pixels.
+```python
+browser.assert_pixels("#detail-content", "filesystem", skip_layouts=["rtl"])
+```
+
+To update pixel tests it's recommended to either start use the tasks container
+to generate specific new pixels or create a draft PR and let the tests run for
```
Comment (martinpitt) [inline test/README.md:136]:
```suggestion
To update pixel tests, locally run the test in the current tasks container, 
or create a draft PR and let the tests run for
```

File: test/README.md [hunk 10]
```diff
@@ -84,10 +84,70 @@ ssh and web.  See the "Helpful tips" section below.
...
+`test/reference-image` and afterwards fetching the new pixels:
```
Comment (martinpitt) [inline test/README.md:137]:
```suggestion
`test/reference-image` and afterwards fetch the new pixels:
```

File: test/README.md [hunk 11]
```diff
@@ -84,10 +84,70 @@ ssh and web.  See the "Helpful tips" section below.
...
+
+```
+./test/common/pixel-tests fetch "https://cockpit-logs.us-east-1.linodeobjects.com/<snip>/log.html
```
Comment (martinpitt) [inline test/README.md:140]: Last time I checked it didn't work with /log.html -- that's the bit that you have to cut off. Maybe that was fixed since then?

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks! Unfortunately suggestions work very poorly with the line breaks. So if you mechanically apply them, please reformat the paragraphs to have a proper width.

File: test/README.md (hunk 11 above)
Comment (mvollmer) [reply test/README.md:140]: Yes, that's fixed.

File: test/README.md [hunk 12]
```diff
@@ -84,10 +84,70 @@ ssh and web.  See the "Helpful tips" section below.
...
+[cockpit/tasks container](https://quay.io/repository/cockpit/tasks) which pins
+the browser and font rendering so repeated runs provide the same pixels. To
+generate new pixels this tasks container should be used as your own browser and
+font rendering software might generate different results. For more information
+read the ["introduction blog post"](https://cockpit-project.org/blog/pixel-testing.html).
+
+The test images are stored in git submodule in the `test/reference` directory and be fetched with:
+
+```sh
+./test/common/pixel-tests fetch
```
Comment (mvollmer) [inline test/README.md:105]: `fetch` does not retrieve things from the Git repo, do you mean `update`?

File: test/README.md (hunk 11 above)
Comment (mvollmer) [inline test/README.md:140]: Unclosed `"`

**Review by mvollmer — CHANGES_REQUESTED** [review]

File: test/README.md (hunk 12 above)
Comment (jelly) [reply test/README.md]: Ah, `pull`!

Comment (jelly) [reply test/README.md]: Oh, TIL, pull is deprecated for `update` good to know :)

File: test/README.md [hunk 13]
```diff
@@ -84,10 +84,71 @@ ssh and web.  See the "Helpful tips" section below.
...
+reference image. if there is a difference, the test fails and a pixel
+difference is shown.
+
+This works as our tests run in the [cockpit/tasks container](https://quay.io/repository/cockpit/tasks)
+which pins the browser and font rendering so repeated runs provide the same
+pixels. To generate new pixels, this tasks container must be used; your own
+browser and font rendering software might generate different results. For more
+information read the ["introduction blog post"](https://cockpit-project.org/blog/pixel-testing.html).
+
+The test images are stored in git submodule in the `test/reference` directory
```
Comment (martinpitt) [inline test/README.md:103]:
```suggestion
The test images are stored in a git submodule in the `test/reference` directory
```

File: test/README.md [hunk 14]
```diff
@@ -84,10 +84,71 @@ ssh and web.  See the "Helpful tips" section below.
...
+./test/common/pixel-tests update
+```
+
+As Cockpit tests under multiple distributions and it is not worth the effort to
+run pixel tests on every supported distribution we only run them for the
+image configured in `test/reference-image`.
+
+Our tests call `Browser.assert_pixels` at interesting and strategic places.
+This assertion method requires at least a CSS selector and an image title.
+Pixel tests are generated in five layouts by default; desktop, medium, mobile,
```
Comment (martinpitt) [inline test/README.md:116]:
```suggestion
Pixel tests are generated in five layouts by default: desktop, medium, mobile,
```

File: test/README.md [hunk 15]
```diff
@@ -84,10 +84,71 @@ ssh and web.  See the "Helpful tips" section below.
...
+dark and rtl.
+
+Take a screenshot of the content in `#detail-content`:
+```python
+browser.assert_pixels("#detail-content", "filesystem")
+```
+
+Take a screenshot the content in `#detail-content` and ignore all elements with
+a class `disk-stats` as they change per test run.
```
Comment (martinpitt) [inline test/README.md:125]:
```suggestion
a class `disk-stats` as they change per test run:
```

File: test/README.md [hunk 16]
```diff
@@ -84,10 +84,71 @@ ssh and web.  See the "Helpful tips" section below.
...
+```python
+browser.assert_pixels("#detail-content", "filesystem", ignore=[".disks-stats"])
+```
+
+Take a screenshot the content in `#detail-content` and skip it for a specific
+layout as it generates unstable pixels.
```
Comment (martinpitt) [inline test/README.md:131]:
```suggestion
layout as it generates unstable pixels:
```

File: test/README.md [hunk 17]
```diff
@@ -84,10 +84,71 @@ ssh and web.  See the "Helpful tips" section below.
...
+
+To update pixel tests, locally run the test in the current tasks container, or
+create a draft PR and let the tests run for `test/reference-image` and
+afterwards fetch the new pixels:
+
+```
+./test/common/pixel-tests fetch "https://cockpit-logs.us-east-1.linodeobjects.com/<snip>/log.html"
+```
+
+And then uploading the new pixel tests and committing the newly generated submodule commit.
```
Comment (martinpitt) [inline test/README.md:144]:
```suggestion
Finally, upload the new pixel tests and commit the newly generated submodule commit:
```

**Review by martinpitt — APPROVED** [review]
Thanks! LGTM by and large. If you push again (maybe @mvollmer has more comments?), please add the trivial editorial fixes, but they are not important.

File: test/README.md [hunk 18]
```diff
@@ -84,10 +84,70 @@ ssh and web.  See the "Helpful tips" section below.
...
+This assertion method requires at least a CSS selector and an image title.
+Pixel tests are generated in five layouts by default: desktop, medium, mobile,
+dark and rtl.
+
+Take a screenshot of the content in `#detail-content`:
+```python
+browser.assert_pixels("#detail-content", "filesystem")
+```
+
+Take a screenshot the content in `#detail-content` and ignore all elements with
```
Comment (mvollmer) [inline test/README.md:124]: `of the content`

File: test/README.md [hunk 19]
```diff
@@ -84,10 +84,70 @@ ssh and web.  See the "Helpful tips" section below.
...
+a class `disk-stats` as they change per test run:
+```python
+browser.assert_pixels("#detail-content", "filesystem", ignore=[".disks-stats"])
+```
+
+Take a screenshot the content in `#detail-content` and skip it for a specific
```
Comment (mvollmer) [inline test/README.md:130]: "of"

**Review by mvollmer — DISMISSED** [review]
Two missing "of", otherwise looks good to me! Thanks!

**Review by martinpitt — APPROVED** [review]
Cheers!

**Pull request merged**